File: C#/LinkedList/SinglyLinkedList.cs
using System;
using System.Collections.Generic;
using Sytem.Text;

namespace LinkedList 
{
	public class SinglyLinkedList<T>
	{
		private class SinglyNode<T> 
		{
			public T Data { get; set; }
			public SinglyNode<T> Next { get; set; }
			
			public SinglyNode(T e) 
			{
				Data = e;
				Next = null;
			}
		}
		
		private SinglyNode<T> _head;
		private SinglyNode<T> _tail;
		
		public T First 
		{
			get { return _head?.Data; }
			set { Count > 0 ? _head.Data = value : AddFirst(value); }
		}
		public T Last 
		{
			get { return _tail?.Data; }
			set { Count > 0 ? _tail.Data = value : AddLast(value); }
		}
		public int Count { get; }
		
		public T this[int i] 
		{
			get { return Get(i); }
			set { Set(i, value); }
		}
		
		public SinglyLinkedList() 
		{
			_head = null;
			_tail = null;
			Count = 0;
		}
		
		public SinglyLinkedList(T[] elems) : this()
		{
			foreach (T e in elems)
			{
				AddLast(e);
			}
		}
		
		public bool IsEmpty()
		{
			return Count == 0;
		}
		
		public T Get(int index) 
		{
			if (index == 0) 
			{
				return First;
			}
			else if (index == Count - 1) 
			{
				return Last;
			}
			else {
				SinglyNode<T> current = _head;
				for (int i = 0; i < index; i++) 
				{
					current = current.Next;
				}
				return current.Data;
			}
		}
		
		public T Set(int index, T e) 
		{
			if (index == 0) {
				T data = First;
				First = e;
				return data;
			}
			else if (index == Count - 1) 
			{
				T data = First;
				Last = e;
				return data;
			}
			else 
			{
				SinglyNode<T> current = _head;
				for (int i = 0; i < index; i++) 
				{
					current = current.Next;
				}
				T data = current.Data;
				current.Data = e;
				return data;
			}
		}
		
		public void AddFirst(T e) 
		{
			if (Count == 0) 
			{
				_head = new SinglyNode<T>(e);
				_tail = _head;
			}
			else 
			{
				SinglyNode<T> temp = new SinglyNode<T>(e);
				temp.Next = _head;
				_head = temp;
			}
			size++;
		}
		
		public void AddLast(T e) 
		{
			if (Count == 0) 
			{
				_tail = new SinglyNode<T>(e);
				_head = _tail;
			}
			else 
			{
				SinglyNode<T> temp = new SinglyNode<T>(e);
				_tail.Next = temp;
				_tail = temp;
			}
			size++;
		}
		
		public void Add(int index, T e) 
		{
			if (index == 0) 
			{
				AddFirst(e);
			}
			else if (index == Count - 1) 
			{
				AddLast(e);
			}
			else 
			{
				SinglyNode<T> previous = _head;
				SinglyNode<T> current = _head.Next;
				for (int i = 0; i < index; i++) 
				{
					previous = current;
					current = current.Next;
				}
				SinglyNode<T> temp = new SinglyNode<T>(e);
				previous.Next = temp;
				temp.Next = current;
				Count++;
			}
		}
		
		public T RemoveFirst() 
		{
			T data = First;
			if (_head == _tail) 
			{
				Clear();
			}
			else 
			{
				_head = _head.Next;
				Count--;
			}
			return data;
		}
		
		public T RemoveLast() 
		{
			T data = Last;
			if (_head == _tail) 
			{
				Clear();
			}
			else 
			{
				SinglyNode<T> current = _head;
				while (current.Next != _tail) 
				{
					current = current.Next;
				}
				current.Next = null;
				_tail = current;
				Count--;
			}
			return data;
		}
		
		public T Remove(int index) 
		{
			if (index == 0) 
			{
				return RemoveFirst();
			}
			else if (index == Count - 1) 
			{
				return RemoveLast();
			}
			else 
			{
				SinglyNode<T> previous = _head;
				SinglyNode<T> current = _head.Next;
				for (int i = 0; i < index; i++) 
				{
					previous = current;
					current = current.Next;
				}
				previous.Next = current.Next;
				Count--;
				return current.Data;
			}
		}
		
		public bool Contains(T e) 
		{
			SinglyNode<T> current = _head;
			while (current != null) 
			{
				if (current.Data.Equals(e)) 
				{
					return true;
				}
				current = current.Next;
			}
			return false;
		}
		
		public int IndexOf(T e) 
		{
			SinglyNode<T> current = _head;
			for (int i = 0; current != null; i++) 
			{
				if (current.Data.Equals(e)) 
				{
					return i;
				}
				current = current.Next;
			}
			return -1;
		}
		
		public bool Erase(T e) 
		{
			int index = IndexOf(e);
			if (index >= 0) 
			{
				Remove(index);
				return true;
			}
			return false;
		}
		
		public void Clear() 
		{
			_head = null;
			_tail = null;
			Count = 0;
		}
	}
}

File: JavaScript/linkedlist/singlylinkedlist.js
class SinglyNode {
	constructor(data) {
		this.data = data;
		this.next = null;
	}
}

class SinglyLinkedList {
	constructor(elems = null) {
		this._head = null;
		this._tail = null;
		this._length = null;
		if (elems) {
			for (let e in elems) {
				this.addLast(e);
			}
		}
	}
	
	get length() {
		return this._length;
	}
	
	isEmpty() {
		return this._length == 0;
	}
	
	get first() {
		if (this._head == null) {
			return null;
		}
		return this._head.data;
	}
	
	get last() {
		if (this._tail == null) {
			return null;
		}
		return this._tail.data;
	}
	
	get(index) {
		if (index == 0) {
			return this.first;
		}
		else if (index == this.length - 1) {
			return this.last;
		}
		else {
			let current = this._head;
			for (let i = 0; i < index; i++) {
				current = current.next;
			}
			return current.data;
		}
	}
	
	set first(e) {
		if (this._length == 0) {
			this.addFirst(e);
		}
		else {
			this._head.data = e;
		}
	}
	
	set last(e) {
		if (this._length == 0) {
			this.addLast(e);
		}
		else {
			this._tail.data = e;
		}
	}
	
	set(index, e) {
		if (index == 0) {
			let data = this.first;
			this.first = e;
			return data;
		}
		else if (index == this._length - 1) {
			let data = this.last;
			this.last = e;
			return data;
		}
		else {
			let current = this._head;
			for (let i = 0; i < index; i++) {
				current = current.next;
			}
			let data = current.data;
			current.data = e;
			return data;
		}
	}
	
	addFirst(e) {
		if (this._length == 0) {
			this._head = new SinglyNode(e);
			this._tail = this._head;
		}
		else {
			let temp = new SinglyNode(e);
			temp.next = this._head;
			this._head = temp;
		}
		this._length++;
	}
	
	addLast(e) {
		if (this._length == 0) {
			this._tail = new SinglyNode(e);
			this._head = this._tail;
		}
		else {
			let temp = new SinglyNode(e);
			this._tail.next = temp;
			this._tail = temp;
		}
		this._length++;
	}
	
	add(index, e) {
		if (index == 0) {
			this.addFirst(e);
		}
		else if (index == this._length - 1) {
			this.addLast(e);
		}
		else {
			let previous = this._head;
			let current = this._head.next;
			for (let i = 0; i < index; i++) {
				previous = current;
				current = current.next;
			}
			let temp = new SinglyNode(e);
			previous.next = temp;
			temp.next = current;
			this._length++;
		}
	}
	
	removeFirst() {
		let data = this.first;
		if (this._head == this._tail) {
			this.clear();
		}
		else {
			this._head = this._head.next;
			this._length--;
		}
		return data;
	}
	
	removeLast() {
		let data = this.last;
		if (this._head == this._tail) {
			this.clear();
		}
		else {
			let current = this._head;
			while (current.next != this._tail) {
				current = current.next;
			}
			current.next = null;
			this._tail = current;
			this._length--;
		}
		return data;
	}
	
	remove(index) {
		if (index == 0) {
			return this.removeFirst();
		}
		else if (index == this._length - 1) {
			return this.removeLast();
		}
		else {
			let previous = this._head;
			let current = this._head.next;
			for (let i = 0; i < index; i++) {
				previous = current;
				current = current.next;
			}
			previous.next = current.next;
			this._length--;
			return current.data;
		}
	}
	
	contains(e) {
		let current = this._head;
		while (current != null) {
			if (current.data == e) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	indexOf(e) {
		let current = this._head;
		for (let i = 0; current != null; i++) {
			if (current.data == e) {
				return i;
			}
			current = current.next;
		}
		return -1;
	}
	
	erase(e) {
		let i = this.indexOf(e)
		if (i >= 0) {
			this.remove(i);
			return true;
		}
		return false;
	}
	
	clear() {
		this._head = null;
		this._tail = null;
		this._length = 0;
	}
}
